packages/core/src/Site/index.js

```diff
@@ -960,61 +991,108 @@ class Site {
    * @param {Array<Page>} pages Pages to be generated
    * @param {ProgressBar} progressBar Progress bar of the overall generation process
-   * @returns {Promise<void>} A Promise that resolves once all pages have been generated
+   * @returns {Promise<boolean>} A Promise that resolves to a boolean which indicates whether the generation
+   * ran to completion
    */
   async generatePagesSequential(pages, progressBar) {
+    const startTime = new Date();
+    let isCompleted = true;
     await utils.sequentialAsyncForEach(pages, async (page) => {
+      if (startTime < this.pageGenerationContext.stopGenerationTimeThreshold) {
+        logger.info('Page generation stopped');
+        logger.debug('Page generation stopped at sequential generation');
+        isCompleted = false;
+        return;
+      }
+
       try {
         await page.generate(this.externalManager);
+        this.toRebuild.delete(FsUtil.removeExtension(page.pageConfig.sourcePath));
+        await this.writeSiteData(false);
         progressBar.tick();
       } catch (err) {
         logger.error(err);
         throw new Error(`Error while generating ${page.sourcePath}`);
       }
     });
+    return isCompleted;
   }
 
   /**
    * Creates the supplied pages' page generation promises at a throttled rate.
    * This is done to avoid pushing too many callbacks into the event loop at once. (#1245)
    * @param {Array<Page>} pages Pages to be generated
    * @param {ProgressBar} progressBar Progress bar of the overall generation process
-   * @return {Promise<void>} A Promise that resolves once all pages have been generated
+   * @returns {Promise<boolean>} A Promise that resolves to a boolean which indicates whether the generation
+   * ran to completion
    */
   generatePagesAsyncThrottled(pages, progressBar) {
     return new Promise((resolve, reject) => {
-      const counter = { numPagesGenerated: 0 };
+      const context = {
+        startTime: new Date(),
+        numPagesGenerated: 0,
+        numPagesToGenerate: pages.length,
+        isCompleted: true,
+      };
 
       // Map pages into array of callbacks for delayed execution
       const pageGenerationQueue = pages.map(page => async () => {
+        // Pre-generate guard to ensure no newly executed callbacks start on stop
+        if (context.startTime < this.pageGenerationContext.stopGenerationTimeThreshold) {
+          if (context.isCompleted) {
+            logger.info('Page generation stopped');
+            logger.debug('Page generation stopped at asynchronous generation');
+            context.isCompleted = false;
+            resolve(false);
+          }
+          return;
+        }
+
         try {
           await page.generate(this.externalManager);
-          Site.generateProgressBarStatus(progressBar, counter, pageGenerationQueue, pages, resolve);
+          this.toRebuild.delete(FsUtil.removeExtension(page.pageConfig.sourcePath));
+          await this.writeSiteData(false);
```

Comment:
let's tighten when this is run (e.g. `regenerateAffectedPages` has repeated calls)

Reply:
Oh right, missed this one. Will do.